### PR TITLE
feat(cursor): per-model usage leaderboard widget

### PR DESCRIPTION
**Pull request by robinebers**
## What

Adds an opt-in **Models** widget for Cursor — a per-model usage leaderboard.

- **Inline**: the top-3 model families by spend, rank-badged and right-aligned (no figures inline) — an at-a-glance read of what you're leaning on.
- **Hover**: expands into a popover listing every model with its spend and token count, spend-sorted.
- **Hover a row**: shows the per-variant cost breakdown (cursorcat-style) — e.g. the `-fast`/`-thinking` splits behind a family — with exact figures. The **Other** row's tooltip lists the families it folded in.

image

## Why

Cursor's usage CSV is already parsed per-row (`CursorUsageCSVRow`) but the model dimension was discarded by the day-bucketed spend aggregation. This surfaces it, reusing pricing/data we already had. Clean family names come straight from the bundled `model_manifest.json` (`family_display_name`), which was previously dead code.

## How

The view layer clones the proven `.chart` / `UsageSparkline` / `UsageTrendDetail` / `TrendHoverState` pattern:

- New `MetricLine.modelBreakdown` case carrying `[ModelUsageEntry]` (name, spend, tokens, `isUnpriced`, per-variant breakdown), threaded through `WidgetData`, `WidgetDataStore.resolve`, the local HTTP API wire format, and the snapshot cache (**key bumped `v5` → `v7`** for the schema change).
- New `CursorModelBreakdown` aggregator — a port of cursorcat's `ModelBreakdownAggregator`, simplified to OpenUsage's imputed-cost path. Groups by model **family** (so `gpt-5.5` + `gpt-5.5-fast` collapse into one "GPT-5.5"), sums spend + tokens, snaps to cents once, sorts by spend.
- New views `ModelLeaderboardRow` (inline) and `ModelLeaderboardDetail` (popover). The popover sizes to content and scrolls past a height cap via **measured content height** (not a row-count guess).
- Low-spend tail (**< 3% of total spend**) folds into a single **"Other"** row, pinned last. Unpriced models are never folded — they stay listed individually with cursorcat's unknown-model warning marker (their cost is unknown, not small).

## Placement (per the new convention)

Per a new **"Adding a Metric"** section added to `AGENTS.md`, the four default-placement decisions were confirmed with the owner rather than chosen silently:

| Decision | Value |
|---|---|
| On / off by default | **Off** (opt-in via Add Widget gallery) |
| Primary / secondary | **Secondary** (behind the provider's expand caret) |
| Pinned | **Not pinnable** (a list has no single tray value) |
| Order | After **Usage Trend**, before the spend tiles |

## Tests

`CursorModelBreakdownTests` covers family collapsing, spend sort, cents snapping, unknown-model handling, the "Other" fold (and its edge cases: single-tail, unpriced-not-folded), and the per-variant breakdown. `CursorSpendTests` adds the data-present and no-data (CSV-failure → "No data") paths. `LayoutStoreTests` updated for the new opt-in descriptor. Full suite: **455 pass**.

## Reviewer notes

- Verified end-to-end against real Cursor data via the local usage API.
- Generalizes to Grok (model attribution already computed) and, with a `--breakdown` flag on the existing ccusage call, to Codex/Claude — left for follow-ups.
- Two deliberate, easily-reversible choices: the per-variant tooltip shows for every row with variants (matching cursorcat), and the 3% threshold is a named constant.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Low Risk**
> Cursor-only, additive metric and UI; no auth or payment changes. Cache key bump forces one refetch on upgrade.
> 
> **Overview**
> Adds an opt-in **Cursor Models** metric that ranks model families by imputed spend over the last 30 days, using the same usage CSV as the spend tiles.
> 
> The change introduces **`MetricLine.modelBreakdown`** plus **`ModelUsageEntry` / `ModelVariantUsage`**, threads them through **`WidgetData`**, **`WidgetDataStore`**, snapshot cache (**v7**), and the local HTTP API (`models` line type). **`CursorModelBreakdown`** aggregates CSV rows by manifest **`family_id`**, cent-snaps totals, folds sub-3% spend into **Other** (with variant hover detail), and **`CursorUsageMapper`** emits the line on successful CSV fetch. UI follows the Usage Trend pattern: **`ModelLeaderboardRow`** (top 3 names inline) and **`ModelLeaderboardDetail`** popover; not pinnable. Default layout keeps **`cursor.models`** off and behind the expand caret. **`AGENTS.md`** documents metric placement rules including **`pinnable: false`** for list/chart widgets.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 304cf3884c932c3ee2ba8023daf51896b24b9e85. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

<h3>Confidence Score: 5/5</h3>

Safe to merge — the new aggregation path is isolated to Cursor, failures degrade to "No data", and no existing behaviour is modified.

The change is well-scoped: it reuses existing CSV data and follows the established chart-row pattern end-to-end. The opt-in placement, non-pinnable flag, and cache version bump are all handled correctly. The test suite covers family collapsing, Other folding, cent-snapping, and the no-CSV-data path. No existing metric paths are modified.

No files require special attention.

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `Sources/OpenUsage/Services/LocalUsageAPI.swift`, line 450-455 ([link](https://github.com/robinebers/openusage/blob/0e4d512cb864ff96a039aac3b413798d10aa77e2/Sources/OpenUsage/Services/LocalUsageAPI.swift#L450-L455)) 

   <a href="#">P2</a> **Wire format omits per-variant breakdown**

   `WireModel` serialises only `name`, `costUSD`, `tokens`, and `unpriced` — the `variants` array that powers the hover tooltip is absent. Any external consumer of the local HTTP API (e.g. a script or future integration) can see the top-level spend per family but cannot reconstruct the per-variant (`gpt-5.5` vs `gpt-5.5-fast`) detail. Is this intentional for the v1 API surface, or should `variants` be included now while the schema is still new?

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: Sources/OpenUsage/Services/LocalUsageAPI.swift
   Line: 450-455

   Comment:
   **Wire format omits per-variant breakdown**

   `WireModel` serialises only `name`, `costUSD`, `tokens`, and `unpriced` — the `variants` array that powers the hover tooltip is absent. Any external consumer of the local HTTP API (e.g. a script or future integration) can see the top-level spend per family but cannot reconstruct the per-variant (`gpt-5.5` vs `gpt-5.5-fast`) detail. Is this intentional for the v1 API surface, or should `variants` be included now while the schema is still new?

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

   <a href="https://app.greptile.com/api/ide/cursor?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20Sources%2FOpenUsage%2FServices%2FLocalUsageAPI.swift%0ALine%3A%20450-455%0A%0AComment%3A%0A**Wire%20format%20omits%20per-variant%20breakdown**%0A%0A%60WireModel%60%20serialises%20only%20%60name%60%2C%20%60costUSD%60%2C%20%60tokens%60%2C%20and%20%60unpriced%60%20%E2%80%94%20the%20%60variants%60%20array%20that%20powers%20the%20hover%20tooltip%20is%20absent.%20Any%20external%20consumer%20of%20the%20local%20HTTP%20API%20%28e.g.%20a%20script%20or%20future%20integration%29%20can%20see%20the%20top-level%20spend%20per%20family%20but%20cannot%20reconstruct%20the%20per-variant%20%28%60gpt-5.5%60%20vs%20%60gpt-5.5-fast%60%29%20detail.%20Is%20this%20intentional%20for%20the%20v1%20API%20surface%2C%20or%20should%20%60variants%60%20be%20included%20now%20while%20the%20schema%20is%20still%20new%3F%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&pr=740&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursorDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursor.svg?v=3">Fix in Cursor</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20Sources%2FOpenUsage%2FServices%2FLocalUsageAPI.swift%0ALine%3A%20450-455%0A%0AComment%3A%0A**Wire%20format%20omits%20per-variant%20breakdown**%0A%0A%60WireModel%60%20serialises%20only%20%60name%60%2C%20%60costUSD%60%2C%20%60tokens%60%2C%20and%20%60unpriced%60%20%E2%80%94%20the%20%60variants%60%20array%20that%20powers%20the%20hover%20tooltip%20is%20absent.%20Any%20external%20consumer%20of%20the%20local%20HTTP%20API%20%28e.g.%20a%20script%20or%20future%20integration%29%20can%20see%20the%20top-level%20spend%20per%20family%20but%20cannot%20reconstruct%20the%20per-variant%20%28%60gpt-5.5%60%20vs%20%60gpt-5.5-fast%60%29%20detail.%20Is%20this%20intentional%20for%20the%20v1%20API%20surface%2C%20or%20should%20%60variants%60%20be%20included%20now%20while%20the%20schema%20is%20still%20new%3F%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=robinebers%2Fopenusage&pr=740&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=3">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22robinebers%2Fopenusage%22%20on%20the%20existing%20branch%20%22claude%2Feager-banach-ecf5a1%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22claude%2Feager-banach-ecf5a1%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20Sources%2FOpenUsage%2FServices%2FLocalUsageAPI.swift%0ALine%3A%20450-455%0A%0AComment%3A%0A**Wire%20format%20omits%20per-variant%20breakdown**%0A%0A%60WireModel%60%20serialises%20only%20%60name%60%2C%20%60costUSD%60%2C%20%60tokens%60%2C%20and%20%60unpriced%60%20%E2%80%94%20the%20%60variants%60%20array%20that%20powers%20the%20hover%20tooltip%20is%20absent.%20Any%20external%20consumer%20of%20the%20local%20HTTP%20API%20%28e.g.%20a%20script%20or%20future%20integration%29%20can%20see%20the%20top-level%20spend%20per%20family%20but%20cannot%20reconstruct%20the%20per-variant%20%28%60gpt-5.5%60%20vs%20%60gpt-5.5-fast%60%29%20detail.%20Is%20this%20intentional%20for%20the%20v1%20API%20surface%2C%20or%20should%20%60variants%60%20be%20included%20now%20while%20the%20schema%20is%20still%20new%3F%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=robinebers%2Fopenusage&pr=740&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=3">Fix in Codex</picture></a>
</details>

<sub>Reviews (3): Last reviewed commit: ["Merge remote-tracking branch &#39;origin/mai..."](https://github.com/robinebers/openusage/commit/304cf3884c932c3ee2ba8023daf51896b24b9e85) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=40037553)</sub>

### CODE REVIEW

**Comment by robinebers** [comment]
This does work for Cursor and could easily be extended for Grok, Claude, and Codex, but I wonder if there is a better way to display this. 

Maybe as a chart or something? Not sure.

File: docs/dashboard.md [hunk 1]
```diff
@@ -21,6 +21,8 @@ Each provider card leads with its **always-shown** metrics. Any metrics you've m
 
 **Usage Trend** (Claude, Codex, Cursor, and Grok) is a small bar chart of the last 30 days of token usage — one bar per day, drawn from the same source as that provider's spend rows (local logs for Claude / Codex / Grok, your usage history for Cursor). **Hover it** for the peak day, the date range, and the source. It's on by default; turn it off or reorder it from Customize like any other metric. It can't be pinned to the menu bar — the strip shows single values, not a chart.
 
+**Models** (Cursor) ranks the models you've used over the last 30 days by spend. The row stays compact — just the top three model names, numbered by rank — as an at-a-glance read of what you're leaning on. **Hover it** to reveal the full list with each model's spend and token count. Models are grouped by family, so a model's faster and reasoning variants count as one. Anything under 5% of your spend is rolled into a single "Other" row to keep the list readable; models Cursor hasn't priced are kept on their own, marked with a small warning. It's off by default — add it from Customize — and, like Usage Trend, it can't be pinned to the menu bar.
```
[the remark this answers is not in the document]
Comment (robinebers) [reply docs/dashboard.md]: Fixed — corrected the doc to "under 3% of your spend" to match `CursorModelBreakdown.tailThresholdFraction = 0.03` (commit 747c700).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: docs/providers/cursor.md [hunk 2]
```diff
@@ -23,6 +24,10 @@ Just be signed into the Cursor app. OpenUsage reads Cursor's local state databas
...
 
+## The models leaderboard
+
+The optional **Models** widget ranks the models you've used over the last 30 days by spend, from the same usage export the spend tiles use. The row stays compact — the top three model names, numbered by rank — and hovering reveals the full list with each model's spend and token count. Models are grouped by family (a model's faster and reasoning variants count as one). Anything under 5% of your spend is grouped into a single "Other" row; a model Cursor hasn't priced is kept on its own, shown with a dash for cost and a small warning marker. Like the spend tiles, the per-model dollars are imputed from token counts at base rates, so they're an estimate.
```
[the remark this answers is not in the document]
Comment (robinebers) [reply docs/providers/cursor.md]: Fixed — corrected the doc to "under 3% of your spend" to match `CursorModelBreakdown.tailThresholdFraction = 0.03` (commit 747c700).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by validatedev** [comment]
@codex review

**Comment by robinebers** [comment]
This was superseeded
